frame/contracts/src/storage/meter.rs

```diff
@@ -533,31 +530,69 @@ impl<T: Config> Ext<T> for ReservingExt {
 
 	fn charge(
 		origin: &T::AccountId,
-		deposit_account: &DepositAccount<T>,
+		contract: &T::AccountId,
 		amount: &DepositOf<T>,
 		terminated: bool,
 	) -> Result<(), DispatchError> {
 		match amount {
 			Deposit::Charge(amount) | Deposit::Refund(amount) if amount.is_zero() => return Ok(()),
 			Deposit::Charge(amount) => {
-				T::Currency::transfer(origin, deposit_account, *amount, Preservation::Preserve)?;
-				Ok(())
+				// This could fail if the `origin` does not have enough liquidity. Ideally, though,
+				// this should have been checked before with `check_limit`.
+				T::Currency::transfer_and_hold(
+					&HoldReason::StorageDepositReserve.into(),
+					origin,
+					contract,
+					*amount,
+					Precision::Exact,
+					Preservation::Preserve,
+					Fortitude::Polite,
+				)?;
+
+				Pallet::<T>::deposit_event(
+					vec![T::Hashing::hash_of(&origin), T::Hashing::hash_of(&contract)],
+					Event::StorageDepositTransferredAndHeld {
+						from: origin.clone(),
+						to: contract.clone(),
+						amount: *amount,
+					},
+				);
 			},
 			Deposit::Refund(amount) => {
-				if terminated {
-					System::<T>::dec_consumers(&deposit_account);
-				}
-				T::Currency::transfer(
-					deposit_account,
+				let transferred = T::Currency::transfer_on_hold(
+					&HoldReason::StorageDepositReserve.into(),
+					contract,
 					origin,
 					*amount,
-					// We can safely make it `Expendable` because our own consumer prevents a
-					// removal.
-					Preservation::Expendable,
+					Precision::BestEffort,
+					Restriction::Free,
+					Fortitude::Polite,
 				)?;
-				Ok(())
+
+				Pallet::<T>::deposit_event(
+					vec![T::Hashing::hash_of(&contract), T::Hashing::hash_of(&origin)],
+					Event::StorageDepositTransferredAndReleased {
+						from: contract.clone(),
+						to: origin.clone(),
+						amount: transferred,
+					},
+				);
+
+				if transferred < *amount {
+					// This should never happen, if it does it means that there is a bug in the
+					// runtime logic. In the rare case this happens we try to refund as much as we
+					// can, thus the `Precision::BestEffort`.
+					log::error!(
+						target: LOG_TARGET,
+						"Failed to repatriate full storage deposit {:?} from contract {:?} to origin {:?}. Transferred {:?}.",
+						amount, contract, origin, transferred,
+					);
+				} else if terminated {
+					System::<T>::dec_providers(&contract)?;
+				}
```

Comment:
I'd never reach this point if `amount.is_zero()`, but termination always carries at least the contract's info storage deposit so we should be good. Still, I'd like some feedback on whether you consider to be safer to check for termination even when `amount` is zero.

Reply:
Yeah let's remove this conditional. If we terminate we always want to remove the provider. Unbalanced providers is pretty bad.

Reply:
We need a way to check if `amount` is zero at some point as `transfer_and_hold` and `transfer_on_hold` don't do it. Otherwise we'd be calling them unnecessarily and we'd be depositing zero-amount events.

I have kept this conditional where it is, but moved the `dec_providers` outside of the `match` statement so that it is always called on a termination process.